### PR TITLE
Add TLS support

### PR DESCRIPTION
**Pull request by mtury**
This is a rework from PR #294 for supporting TLS in Scapy.

It relies on the `cryptography` library which was adopted with PR #371.
The `pycrypto` and `ecdsa` modules are no longer needed.

Documentation taken from the `__init__.py`:

Main features:
- X.509 certificates parsing/building.
- RSA & ECDSA keys sign/verify methods.
- TLS records and sublayers (handshake...) parsing/building. Works with versions SSLv3, TLS 1.0, 1.1 and 1.2. This may be enhanced by a TLS context. For instance, if scapy reads a ServerHello with version TLS 1.2 and a cipher suite using AES, it will assume the presence of IVs prepending the data. See test/tls.uts for real examples.
- TLS encryption/decryption capabilities with the usual ciphersuites. Once again, the TLS context enables scapy to transparently send/receive protected data if it learnt the session secrets. Note that if scapy acts as one side of the handshake (e.g. reads all server-related packets and builds all client-related packets), it will indeed compute the session secrets.
- TLS client & server basic automatons, provided for testing and tweaking purposes. These make for a very primitive TLS stack.
- Additionally, a basic test PKI (key + certificate for a CA, a client and a server) is provided in tls/examples/pki_test.

Unit tests:
- Various cryptography checks.
- Reading a TLS handshake between a Firefox client and a GitHub server.
- Test our TLS server against s_client with different cipher suites.
- Test our TLS client against our TLS server (s_server is unscriptable).

TODO list (may it be carved away by good souls):
- Features to add (or wait for) in the `cryptography` library:
        - no limitation on FFDH generator size;
          (remove line 88 in cryptography/hazmat/primitives/asymmetric/dh.py)
        - CCM and CHACHA20-POLY1305 ciphers;
        - ECDH curves (x25519 and x448) from RFC 7748;
        - FFDH groups from RFC 7919;
        - the so-called 'tls' hash used with SSLv3 and TLS 1.0;
        - the simple DES algorithm;
        - the compressed EC point format.

- About the automatons:
        - Enrich the automatons. The client should be able to receive data at any time, and to send as much data as wanted from stdin (for now, only one predefined data message may be sent following the handshake). The server should stay online even after the first client leaves. Then we could look at more complicated behaviours like renegotiation and resumption. We might get some help from [tintinweb/scapy-ssl_tls](https://github.com/tintinweb/scapy-ssl_tls).
        - Add some examples which illustrate how the automatons could be used. Typically, we could showcase this with Heartbleed.
        - Split up parts of the automaton, e.g. when our server builds the ServerHello, Certificate, ServerKeyExchange and ServerHelloDone in the same should_REPLY_TO_CH method.
        - Make the automatons tests more robust and less consuming.
        - Allow the server to store both one RSA key and one ECDSA key, and select the right one to use according to the ClientHello suites.
        - Find a way to shutdown the automatons sockets properly without simultaneously breaking the unit tests.

- Miscellaneous:
        - Implement TLS 1.3 structures and automatons. :D
        - Implement SSLv2 structures and automatons. xD
        - Mostly unused features : DSS, fixed DH, SRP, IDEA, char2 curves...
        - Check FFDH and ECDH parameters at SKE/CKE reception.
        - Go through the kx_algs and see what may be commented out.
        - Define several Certificate Transparency objects.
        - Enhance PSK and session ticket support.

### CODE REVIEW

File: .travis/install.sh [hunk 1]
```diff
@@ -10,9 +10,9 @@ fi
 
 if python --version 2>&1 | grep -q PyPy; then
   # cryptography requires PyPy >= 2.6, Travis CI uses 2.5.0
-  $SCAPY_SUDO pip install $PIP_INSTALL_FLAGS ecdsa mock
+  $SCAPY_SUDO pip install $PIP_INSTALL_FLAGS mock
 else
-  $SCAPY_SUDO pip install $PIP_INSTALL_FLAGS cryptography ecdsa mock
+  $SCAPY_SUDO pip install $PIP_INSTALL_FLAGS cryptography mock
```
Comment (p-l-) [inline .travis/install.sh]: The version of Pypy used in Travis is not recent enough for cryptography. Please remove that change.

File: scapy/config.py [hunk 2]
```diff
@@ -296,36 +296,69 @@ def _prompt_changer(attr,val):
...
-warning_threshold : how much time between warnings from the same place
-ASN1_default_codec: Codec used by default for ASN1 objects
-mib      : holds MIB direct access dictionary
-resolve   : holds list of fields for which resolution should be done
-noenum    : holds list of enum fields for which conversion to string should NOT be done
-AS_resolver: choose the AS resolver class to use
-extensions_paths: path or list of paths where extensions are to be looked for
-contribs: a dict which can be used by contrib layers to store local configuration
-"""
+    """
```
Comment (p-l-) [inline scapy/config.py]: Please leave this as it was, I think it's easier to read.

File: test/tls.uts [hunk 3]
```diff
@@ -0,0 +1,955 @@
+% Tests for TLS module
+# 
+# Try me with :
+# bash test/run_tests -t test/tls.uts -F
+
```
Comment (p-l-) [inline test/tls.uts]: Please add `~ crypto` here, so that Pypy tests ignore this file.

**Comment by gpotter2** [comment]
@mtury Well about your TODO list:
automaton is already not working under windows, see https://github.com/secdev/scapy/issues/451. If you want to have a look at that too...

File: .travis/install.sh (hunk 1 above)
Comment (mtury) [inline .travis/install.sh]: I'm not sure about what you're asking here. I only removed the `ecdsa`module, which is no longer used by scapy.

Comment (p-l-) [inline .travis/install.sh]: Sorry I misunderstood your change.

**Comment by p-l-** [comment]
@mtury could you rebase against current master? That should help us understand what's wrong with coverage.

**Comment by mtury** [comment]
FYI `python-coverage` does not seem to handle our automata correctly. I could run a TLS server locally but it failed to show execution in the `TLSServerAutomaton` methods. This explains in particular the poor coverage for `pkcs1.py` and `automaton.py`. Not sure what I could do here.

File: test/run_tests_tls_netaccess [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  msg=$1
+  suite=$2
+  version=$3
+  EXIT_CODE=test_tls_server_tmp_code
+  OSSL_STDO=test_tls_server_tmp_openssl_stdout
+  OSSL_STDE=test_tls_server_tmp_openssl_stderr
+  SCAPY_STDE=test_tls_server_tmp_scapy_stderr
+
+  echo "1" > $EXIT_CODE
+  ($TRAVIS_SUDO python tls_test_server.py "$msg" 2>$SCAPY_STDE; echo $? > $EXIT_CODE) &
```
Comment (gpotter2) [inline test/run_tests_tls_netaccess]: Have you tried to replace this by something like used here:
https://github.com/secdev/scapy/blob/0c9b9089abffcffc788cc525ac7f2905c840d49d/appveyor.yml#L27
To make sure that codecoverage is used ?

**Comment by gpotter2** [comment]
Apart from that, I don't really like having a "examples" folder inside scapy's main code...

File: test/run_tests_tls_netaccess (hunk 4 above)
Comment (mtury) [inline test/run_tests_tls_netaccess]: I tried to but results were not conclusive. In the end I ran `coverage` directly from my shell on `tls_test_server.py`, and it confirmed that (apart from the `__init__`) the code inside the automaton was not covered.

**Comment by mtury** [comment]
You're right about that, I'll move it soon to the `test` folder!

**Comment by gpotter2** [comment]
@mtury I got a question: why are you using `run_tests_tls_........` ?
It would be better to use a .uts file.

For the integration of your tests, 
`or.system("sudo openssl s_client -cipher...")`
will work, and run the command as root.

**Comment by mtury** [comment]
I guess it would be possible to rewrite the bash script as a UTS file, but I don't think it's worth the trouble. Would there be strong benefits from this?

Aside from that, while speaking of `test/` rules, doesn't it seem right to bring to it the UTS scripts from `scapy/contrib/`?

**Comment by gpotter2** [comment]
The advantages to use a UTS file is that UTscapy will handle them, which is nice. Then, that will surely fix the codecoverage problems.

The scapy/contrib folder is a bit on his own... I guess that's why the tests were left with their .py files

**Comment by gpotter2** [comment]
You've done a great job, here's an advice how to do the tests:

First, change the implementation of tls_test_server.py to something like:

```
def run_tls_test(expected_data=None):
    with captured_output() as (out, err):
    t = TLSServerAutomaton(mycert=basedir+'/scapy/layers/tls/examples/pki_test/srv_cert.pem',
                           mykey=basedir+'/scapy/layers/tls/examples/pki_test/srv_key.pem')
    t.run()
    check_output_for_data(out, err, expected_data)
```

Then the tests won't be that hard, and you don't even have to take care of the error/out streams:

```
% Regression tests for TLS

# More informations at http://www.secdev.org/projects/UTscapy/

############
############
+ Test TLS server automaton

= Define util functions

import tls_test_server, time, thread, subprocess

def test_tls_server(msg, suite="", version=""):
    def start_server():
        run_tls_test(msg)
    thread.start_new_thread(start_server, ())
    time.sleep(2)
    p = subprocess.Popen("sudo openssl s_client -cipher " + suite + " " + version)
    out, err = p.communicate()
    if err:
        print err
        raise RuntimeError("OpenSSL returned with error code")
    assert p.returncode == 0

= Your first test

test_tls_server....
```

What do you think ?

**Comment by mtury** [comment]
Thanks for the help @gpotter2. Unfortunately I find myself unable to work out a UTS script... I'm afraid (though unsure) that automatons do not accommodate with threads.

Since the codecov warning should be ignored (as long as automatons are not supported), I believe the PR is as ready as can be!

**Comment by gpotter2** [comment]
@mtury 
 Ohhhh, i just remembered something:
In https://github.com/secdev/scapy/pull/496 i had found out that coverage did NOT work with the `thread` module, which I had to replace with `threading`.
It's not merged yet (@p-l- It should be ready to merge), so can you try again with it ?

File: scapy/layers/tls/__init__.py [hunk 5]
```diff
@@ -1,24 +1,121 @@
...
+
+    - Additionally, a basic test PKI (key + certificate for a CA, a client and
+      a server) is provided in tls/examples/pki_test.
+
+
+Unit tests:
+
+    - Various cryptography checks.
+
+    - Reading a TLS handshake between a Firefox client and a GitHub server.
```
Comment (gpotter2) [inline scapy/layers/tls/__init__.py]: In the secdev team, they love calling "secdev.org" rather than any other servers :)

**Comment by mtury** [comment]
Indeed I tried with `threading` after failing with `thread`, but to no avail. :/

Regarding the secdev.org query, I'll be sure to favor it in the future! Many tests of the UTS follow on from the github domain name, and it's a passive packet analysis anyway, so I don't think it's worth bothering to rewrite it. Furthermore, the certificate on secdev.org isn't much of an example to follow. :)

I squashed some commits in order to make the history clearer before they are merged.

File: test/tls/travis_test_server.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+@contextmanager
+def captured_output():
+    new_out, new_err = StringIO(), StringIO()
+    old_out, old_err = sys.stdout, sys.stderr
+    try:
+        sys.stdout, sys.stderr = new_out, new_err
+        yield sys.stdout, sys.stderr
+    finally:
+        sys.stdout, sys.stderr = old_out, old_err
```
Comment (gpotter2) [inline test/tls/travis_test_server.py]: I do not like that part :/ Parsing the output is a bit dirty

**Comment by gpotter2** [comment]
@mtury Because i'm santa claus, i made you a working (at least for me) version of UTS tests :)
https://github.com/gpotter2/scapy/commit/a12931c9c255a75b7d03225c09b151752c88d9b4

**Comment by guedou** [comment]
@gpotter2 I am fine with the current PR. I think that we must merge it as it is, then modify details if needed.

**Comment by mtury** [comment]
@gpotter2 Thanks for your help. I also think that you could open another PR once the current one is accepted, which I'll be sure to review!

**Comment by gpotter2** [comment]
I understand, anyway it's harder than I thought it would be... (it's hard to emulate exit() behavior without closing everything)

**Comment by gpotter2** [comment]
@mtury Hey i'm occuring this, is it a bug:

```
......scapy\scapy\layers\tls\keyexchange.py", line 632, in post_dissection
    self.tls_session.server_tmp_rsa_key = PubKeyRSA((e, m, mLen))
NameError: global name 'PubKeyRSA' is not defined
```
when running the SSLv3 tests

**Pull request merged**